Add tests for shellsyntax for /bin/rc

sturivny opened the pull request:
@dcantrell Some tests are failed. 
Please, take a look.
Thanks.

dcantrell commented:
This one has merge conflicts for Dockerfile.test.

Also it looks like the tests failed due to rc syntax.

sturivny commented:
```
[sturivny@tempo]$ cat btest.rc
#!/bin/rc
who >user.names
who >>user.names
wc <file

echo [a-f]*.c
who | wc
who; date
vc *.c &)

mk && v.out /*/bin/fb/*
echo hully^gully

{sleep 3600;echo 'Time''s up!'}&

for(i in *.c) if(cpp $i >/tmp/$i) vc /tmp/$i

fn g {
    grep $1 *.[hcyl]
}}

vc junk.c >2=1] >junk.out
[sturivny@tempo]$ rc -n btest.rc
rc: line 10: syntax error near '='
```

dcantrell commented:
That looks good as an invalid /bin/rc script, but the PR includes all the previous tcsh code as well.  Including invalid_tcsh that is actually valid.  I think this PR needs to be fixed up.

sturivny commented:
Ah. Going to fix it.

dcantrell commented:
Were you going to rebase this against master and fix up the failures?  Need any help with that?  I would like to merge the test cases for rc scripts.

sturivny commented:
@dcantrell 
Rebased onto `master` and updated PR with your suggestions.
Got 6 failed tests:
```
runTest (test_shellsyntax.RcMalformedCompareKoji) ... FAIL
runTest (test_shellsyntax.RcMalformedCompareRPMs) ... FAIL
runTest (test_shellsyntax.RcMalformedKoji) ... FAIL
runTest (test_shellsyntax.RcMalformedRPM) ... FAIL
runTest (test_shellsyntax.RcWellMalformedCompareKoji) ... FAIL
runTest (test_shellsyntax.RcWellMalformedCompareRPMs) ... FAIL
```

Invalid rc file:
```
[sturivny@tempo]$ cat rc_inv.rc 
#!/bin/rc
who >user.names
who >>user.names
wc <file
echo [a-f]*.c
who | wc
who; date
vc *.c &)
mk && v.out /*/bin/fb/*
echo hully^gully
{sleep 3600;echo 'Time''s up!'}&
for(i in *.c) if(cpp $i >/tmp/$i) vc /tmp/$i
fn g {
    grep $1 *.[hcyl]
}}
vc junk.c >2=1] >junk.out
```

Check:
```
[sturivny@tempo]$ rc -n rc_inv.rc 
rc: line 8: syntax error near ')'
rc: line 15: syntax error near '}'
rc: line 16: syntax error near '='
```

May you take a look?
Thanks :)

dcantrell commented:
I think I've found the problem.  /bin/rc when run with -n does not exit with non-zero if there is a syntax error.  It just prints the syntax errors to stderr but returns 0.  This needs some special handling in rpminspect.  It appears to be the only shell that does this.

dcantrell commented:
I pushed an update to the inspect_shellsyntax.c code that hanldes the way /bin/rc returns syntax errors.  I just ran the test suite with your /bin/rc tests and everything passes.  Can you update your branch with this change from master and have it retest again in Travis-CI?  You don't need to make any code changes.

sturivny commented:
Sure

On Fri, Apr 17, 2020 at 4:24 PM David Cantrell <notifications@github.com>
wrote:

> I pushed an update to the inspect_shellsyntax.c code that hanldes the way
> /bin/rc returns syntax errors. I just ran the test suite with your /bin/rc
> tests and everything passes. Can you update your branch with this change
> from master and have it retest again in Travis-CI? You don't need to make
> any code changes.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rpminspect/rpminspect/pull/118#issuecomment-615273751>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVZAZ6PYMXY5IGOM72S6SLRNBRCNANCNFSM4LWRNBKA>
> .
>


-- 
Thanks a lot in advance.
Kind Regards, Serhii Turivnyi

Mobile:  +420 608 789 719

dcantrell commented:
Thanks!
